samza-core/src/main/scala/org/apache/samza/diagnostics/DiagnosticsStreamMessage.java

```diff
@@ -0,0 +1,305 @@
+        new DiagnosticsStreamMessage(metricsSnapshot.getHeader().getJobName(), metricsSnapshot.getHeader().getJobId(),
+            metricsSnapshot.getHeader().getContainerName(), metricsSnapshot.getHeader().getExecEnvironmentContainerId(),
+            metricsSnapshot.getHeader().getVersion(), metricsSnapshot.getHeader().getSamzaVersion(),
+            metricsSnapshot.getHeader().getHost(), metricsSnapshot.getHeader().getTime(),
+            metricsSnapshot.getHeader().getResetTime());
+
+    if (metricsSnapshot.getMetrics().getAsMap().containsKey(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)) {
+      diagnosticsStreamMessage.addContainerNumCores((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_NUM_CORES_METRIC_NAME));
+
+      diagnosticsStreamMessage.addContainerMb((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_MB_METRIC_NAME));
+
+      diagnosticsStreamMessage.addNumStoresWithChangelog((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_NUM_STORES_WITH_CHANGELOG_METRIC_NAME));
+
+      diagnosticsStreamMessage.addContainerModels(deserializeContainerModelMap(
+          (Map<String, Object>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+              .get(CONTAINER_MODELS_METRIC_NAME)));
+
+      diagnosticsStreamMessage.addProcessorStopEvents(
+          (List<DiagnosticsManager.ProcessorStopEvent>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+              .get(STOP_EVENT_LIST_METRIC_NAME));
+    }
+
+    if (metricsSnapshot.getMetrics().getAsMap().containsKey(SAMZACONTAINER_METRICS_GROUP_NAME)) {
+      diagnosticsStreamMessage.addDiagnosticsExceptionEvents(
+          (Collection<DiagnosticsExceptionEvent>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(SAMZACONTAINER_METRICS_GROUP_NAME)
+              .get(EXCEPTION_LIST_METRIC_NAME));
+    }
+
+    return diagnosticsStreamMessage;
+  }
+
+  /**
+   * Helper method to use {@link SamzaObjectMapper} to serialize {@link ContainerModel}s.
+   * {@link SamzaObjectMapper} provides several conventions and optimizations for serializing containerModels.
+   * @param containerModelMap map of container models to serialize.
+   * @return
+   */
+  private static Map<String, String> serializeContainerModelMap(Map<String, ContainerModel> containerModelMap) {
+    Map<String, String> serializedContainerModelMap = new HashMap<>();
+    ObjectMapper samzaObjectMapper = SamzaObjectMapper.getObjectMapper();
+    String serializedContainerModel;
+
+    for (Map.Entry<String, ContainerModel> containerModelEntry : containerModelMap.entrySet()) {
+      serializedContainerModel = "";
```

Comment:
This empty string might get put into the map. Will Jackson read that correctly?

Reply:
yes, that gets handled; so that we can get the container count, even if the container-model doesnt get serialized for any reason.

Reply:
Looking at `deserializeContainerModelMap`, an entry that can't be deserialized seems to get skipped. So it doesn't look like you would be able to get the correct container count if some or all did not get put into the map.

Reply:
agree, tried simplifying this.